File: binding/src/libqfileselector.cpp
#include <QFileSelector>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QUrl>
#include <qfileselector.h>
#include "libqfileselector.h"
#include "libqfileselector.hxx"

QFileSelector* QFileSelector_new() {
	 return new QFileSelector();
}

QFileSelector* QFileSelector_new2(QObject* parent) {
	 return new QFileSelector(parent);
}

libqt_string QFileSelector_Tr(const char* s) {
	QString _ret = QFileSelector::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QFileSelector_Select(const QFileSelector* self, const libqt_string filePath) {
	QString _ret = self->select(QString::fromUtf8(filePath.data, filePath.len));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QUrl* QFileSelector_Select2(const QFileSelector* self, const QUrl* filePath) {
	return new QUrl(self->select(*filePath));
}

libqt_list QFileSelector_ExtraSelectors(const QFileSelector* self) {
	auto _ret = self->extraSelectors();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

void QFileSelector_SetExtraSelectors(QFileSelector* self, const libqt_list list) {
	self->setExtraSelectors(QList<QString>());
}

libqt_list QFileSelector_AllSelectors(const QFileSelector* self) {
	auto _ret = self->allSelectors();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

libqt_string QFileSelector_Tr2(const char* s, const char* c) {
	QString _ret = QFileSelector::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QFileSelector_Tr3(const char* s, const char* c, int n) {
	QString _ret = QFileSelector::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QFileSelector_Delete(QFileSelector* self) {
    delete self;
}

File: binding/include/libqfileselector.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qfileselector.html
pub const qfileselector = struct {

    /// New constructs a new QFileSelector object.
    pub fn New() QtC.QFileSelector {
        return qtc.QFileSelector_new();
    }


    /// New2 constructs a new QFileSelector object.
    pub fn New2(parent: ?*anyopaque) QtC.QFileSelector {
        return qtc.QFileSelector_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFileSelector_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfileselector.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Select(self: ?*anyopaque, filePath: []const u8, allocator: std.mem.Allocator) []const u8 {
        const filePath_str = qtc.libqt_string{
    .len = filePath.len,
    .data = filePath.ptr,
};
const _str = qtc.QFileSelector_Select(@ptrCast(self), filePath_str);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfileselector.Select: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Select2(self: ?*anyopaque, filePath: ?*anyopaque) QtC.QUrl {
        return qtc.QFileSelector_Select2(@ptrCast(self), @ptrCast(filePath));
    }

    pub fn ExtraSelectors(self: ?*anyopaque, ) []const u8 {
        return qtc.QFileSelector_ExtraSelectors(@ptrCast(self));
    }

    pub fn SetExtraSelectors(self: ?*anyopaque, list: []const u8) void {
        qtc.QFileSelector_SetExtraSelectors(@ptrCast(self), @ptrCast(list));
    }

    pub fn AllSelectors(self: ?*anyopaque, ) []const u8 {
        return qtc.QFileSelector_AllSelectors(@ptrCast(self));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFileSelector_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfileselector.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFileSelector_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfileselector.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QFileSelector_Delete(@ptrCast(self));
    }
};
